components/script/dom/customelementregistry.rs

```diff
@@ -456,6 +456,196 @@ impl CustomElementDefinition {
+                    (definition.callbacks.adopted_callback.clone(), args)
+                },
+                CallbackReaction::AttributeChanged(local_name, old_val, val, namespace) => {
+                    // Step 4
+                    if !definition.observed_attributes.iter().any(|attr| *attr == *local_name) {
+                        return;
+                    }
+
+                    let cx = element.global().get_cx();
+
+                    let local_name = DOMString::from(&*local_name);
+                    rooted!(in(cx) let mut name_value = UndefinedValue());
+                    unsafe { local_name.to_jsval(cx, name_value.handle_mut()); }
+
+                    rooted!(in(cx) let mut old_value = NullValue());
+                    if let Some(old_val) = old_val {
+                        unsafe { old_val.to_jsval(cx, old_value.handle_mut()); }
+                    }
+
+                    rooted!(in(cx) let mut value = NullValue());
+                    if let Some(val) = val {
+                        unsafe { val.to_jsval(cx, value.handle_mut()); }
+                    }
+
+                    let namespace = DOMString::from(&*namespace);
+                    rooted!(in(cx) let mut namespace_value = UndefinedValue());
+                    unsafe { namespace.to_jsval(cx, namespace_value.handle_mut()); }
+
+                    let mut args = Vec::with_capacity(4);
+                    args.push(Heap::default());
+                    args.push(Heap::default());
+                    args.push(Heap::default());
+                    args.push(Heap::default());
+                    args.get_mut(0).unwrap().set(name_value.get());
+                    args.get_mut(1).unwrap().set(old_value.get());
+                    args.get_mut(2).unwrap().set(value.get());
+                    args.get_mut(3).unwrap().set(namespace_value.get());
+
+                    (definition.callbacks.attribute_changed_callback.clone(), args)
+                },
+            };
+
+            // Step 3
+            let callback = match callback {
+                Some(callback) => callback,
+                None => return,
+            };
+
+            // Step 5
+            element.push_reaction(CustomElementReaction::Callback(callback, args.into_boxed_slice()));
+
+            // Step 6
+            self.enqueue_element(element);
+        }
+    }
+}
+
+/// https://html.spec.whatwg.org/multipage/#element-queue
+#[derive(HeapSizeOf, JSTraceable)]
+#[must_root]
+struct ElementQueue {
+    queue: DOMRefCell<Vec<JS<Element>>>,
+}
+
+impl ElementQueue {
+    fn new() -> ElementQueue {
+        ElementQueue {
+            queue: Default::default(),
+        }
+    }
+
+    /// https://html.spec.whatwg.org/multipage/#invoke-custom-element-reactions
+    fn invoke_reactions(&self) {
+        // Steps 1-2
+        for element in self.queue.borrow().iter() {
+            element.invoke_reactions()
```

Comment:
Oh actually in the second case, an element can be appended to the backup element queue is being invoked. The flag check occurs after the element is appended to the queue. This is what causes the crash in the test expectations.

I want to change it so the queue is not borrowed for the duration of the loop. Instead if element are popped and processed, new element can be appended to the queue. But I always run into unrooted must root errors. any ideas?

```
while let Some(element) = self.queue.borrow_mut().pop_front().map(|e| Root::from_ref(&*e)) {
    element.invoke_reactions()
}
```